src/octree/batch_iterator.rs

```diff
@@ -107,66 +103,71 @@ where
-            return self.callback();
+impl<F> Iterator for BatchIterator<F>
+where
+    F: Fn(&Point) -> bool,
+{
+    type Item = PointData;
+    fn next(&mut self) -> Option<Self::Item> {
+        let max_points = self.position.capacity();
+        for _num_point in 0..max_points {
+            match self.point_iterator.next() {
+                Some(point) => {
+                    self.push_point(point);
+                }
+                None => return self.extract_pointdata(),
+            }
         }
-        Ok(())
+        self.extract_pointdata()
     }
 }
 
-/// Iterator on point batches
-pub struct BatchIterator<'a> {
-    octree: &'a Octree,
-    point_location: &'a PointQuery,
+/// compute a function while iterating on a batch of points
+pub fn try_for_each_point_batch<'a, G>(
+    octree: &'a octree::Octree,
+    point_query: &'a PointQuery,
     batch_size: usize,
-}
-
-impl<'a> BatchIterator<'a> {
-    pub fn new(
-        octree: &'a octree::Octree,
-        point_location: &'a PointQuery,
-        batch_size: usize,
-    ) -> Self {
-        BatchIterator {
-            octree,
-            point_location,
-            batch_size,
+    mut func: G,
+) -> Result<()>
+where
+    G: FnMut(PointData) -> Result<()>,
+{
+    let mut point_batch_vec = get_point_data_iterator(octree, point_query, batch_size);
+    // operate on nodes
+    while let Some(mut point_batch) = point_batch_vec.pop_front() {
+        while let Some(pointdata) = point_batch.next() {
+            if let Err(err) = func(pointdata) {
+                return Err(err);
+            }
         }
     }
 
-    /// compute a function while iterating on a batch of points
-    pub fn try_for_each_batch<F>(&mut self, mut func: F) -> Result<()>
-    where
-        F: FnMut(PointData) -> Result<()>,
-    {
-        //TODO(catevita): mutable function parallelization
-        let local_from_global = self
-            .point_location
-            .global_from_local
-            .clone()
-            .map(|t| t.inverse());
-        let mut point_stream = PointStream::new(self.batch_size, local_from_global, &mut func);
-        // nodes iterator: retrieve nodes
-        let node_id_iterator = self.octree.nodes_in_location(self.point_location);
-        // operate on nodes
-        for node_id in node_id_iterator {
-            let point_iterator = self.octree.points_in_node(self.point_location, node_id);
-            for point in point_iterator {
-                point_stream.push_point_and_callback(point)?;
-            }
-        }
-        // TODO(catevita): return point data through mpsc channel
-        // TODO(catevita): apply mut function to received data
+    Ok(())
+}
 
-        Ok(())
-    }
+pub fn get_point_data_iterator<'a>(
+    octree: &'a octree::Octree,
+    point_query: &'a PointQuery,
+    batch_size: usize,
+) -> VecDeque<Box<BatchIterator<impl Fn(&Point) -> bool>>> {
+    let local_from_global = point_query.global_from_local.clone().map(|t| t.inverse());
```

Comment:
Nit: I think you can avoid the clone with `as_ref()`

Reply:
done